### ISSUE TITLE
feature: support approx_count_distinct

### ISSUE THREAD

**Issue by fuyufjh** [singularity-data/risingwave#2727]
**Is your feature request related to a problem? Please describe.**

Approximate Count Distinct (`approx_count_distinct`) is a function widely supported in may database/streaming system. It use HyperLogLog algorithm to do count distinct with very little cost, compared with a precise `COUNT(DISTINCT)`.

https://docs.microsoft.com/en-us/sql/t-sql/functions/approx-count-distinct-transact-sql?view=sql-server-ver15

**Describe the solution you'd like**

See above.

**Describe alternatives you've considered**

N/A

**Additional context**

Did HyperLogLog supports delete/update?

**Comment by liurenjie1024** [comment]
Yes, that's my question, how to support it in streaming with delete/update?

**Comment by Graphcalibur** [comment]
I'd like to try implementing this feature. Are there any recommendations for sources/readings on how the HyperLogLog algorithm is implemented in other systems? @fuyufjh

**Comment by lmatz** [comment]
> I'd like to try implementing this feature. Are there any recommendations for sources/readings on how the HyperLogLog algorithm is implemented in other systems? @fuyufjh

Redis, Presto

**Comment by fuyufjh** [comment]
> Did HyperLogLog supports delete/update?

Today I came up with an idea about how to support deletes & updates in HyperLogLog. It will use more space than the original non-updatable version, but seems still acceptable.

First, let's revisit the internals of the original non-deletable HyperLogLog. It needs a list of 6-bit integers to store the max suffix-0 count of each bucket. In Redis, the bucket number is set to 16384; since the count of zero must < 64, 6 bits are enough.

```
// Pseudo code
sturct HyperLogLog {
    data: uint6[16384]
}
``` 

To make it updatable, instead of storing a single 6-bit integer for each bucket, we have to maintain a table of "how many numbers are with N suffix-0" for each bucket.

```
// Pseudo code
struct Bucket {
   count: usize[64]
}
sturct HyperLogLog {
   data: Bucket[16384]
}
```

Therefore, once a value is deleted, we could decrease the count of it from its corresponding bucket. 

Furthermore, the `usize` in `Bucket` is too luxury. According to probability, the size of 2nd bucket is expected to be half of the 1st, and the size of 3rd bucket is expected to be half of the 2nd... and so on. Thus, we may simply allocate various size for these counts, for example,

```
// Pseudo code
struct Bucket {
   count_0_to_4: uint32[4]
   count_4_to_20: uint16[16]
   count_20_to_64: uint8[44]
}
```

That would be enough, I think.

**Comment by Graphcalibur** [comment]
That seems like a great way to handle deletes/updates! I just have two questions:

* What's the expected size of data that will be streamed and counted using `approx_distinct_count`
* Is there any worry that the user deleting nonexistent elements can cause the value of one of the buckets to underflow?

**Comment by fuyufjh** [comment]
> What's the expected size of data that will be streamed and counted using approx_distinct_count

It's hard to say. Can very from a few to billions.

> Is there any worry that the user deleting nonexistent elements can cause the value of one of the buckets to underflow?

This can not happend.
